From 968971bd7fed9ee7391f3b3223124ee8b5f0d147 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 18 Aug 2014 19:56:36 -0400
Subject: [scsi] qla2xxx: Check for peg alive counter and clear any outstanding mailbox command

Message-id: <1408391817-6389-49-git-send-email-cdupuis@redhat.com>
Patchwork-id: 87941
O-Subject: [PATCH RHEL 7.1 e-stor 48/69] qla2xxx: Check for peg alive counter and clear any outstanding mailbox command.
Bugzilla: 1089346
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Giridhar Malavali <giridhar.malavali@qlogic.com>

scsi commit id 7012532dc8cdbcfb0d13d03ccc2ec18416504f29

Signed-off-by: Giridhar Malavali <giridhar.malavali@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_nx2.c b/drivers/scsi/qla2xxx/qla_nx2.c
index 063d574..ce15146 100644
--- a/drivers/scsi/qla2xxx/qla_nx2.c
+++ b/drivers/scsi/qla2xxx/qla_nx2.c
@@ -2124,6 +2124,13 @@ qla8044_watchdog(struct scsi_qla_host *vha)
      test_bit(FCOE_CTX_RESET_NEEDED, &vha->dpc_flags))) {
   dev_state = qla8044_rd_direct(vha, QLA8044_CRB_DEV_STATE_INDEX);
 
+  if (qla8044_check_fw_alive(vha)) {
+   ha->flags.isp82xx_fw_hung = 1;
+   ql_log(ql_log_warn, vha, 0xb10a,
+       "Firmware hung.\n");
+   qla82xx_clear_pending_mbx(vha);
+  }
+
   if (qla8044_check_temp(vha)) {
    set_bit(ISP_UNRECOVERABLE, &vha->dpc_flags);
    ha->flags.isp82xx_fw_hung = 1;
@@ -2144,7 +2151,7 @@ qla8044_watchdog(struct scsi_qla_host *vha)
    qla2xxx_wake_dpc(vha);
   } else  {
    /* Check firmware health */
-   if (qla8044_check_fw_alive(vha)) {
+   if (ha->flags.isp82xx_fw_hung) {
     halt_status = qla8044_rd_direct(vha,
      QLA8044_PEG_HALT_STATUS1_INDEX);
     if (halt_status &
@@ -2180,12 +2187,8 @@ qla8044_watchdog(struct scsi_qla_host *vha)
           __func__);
       set_bit(ISP_ABORT_NEEDED,
           &vha->dpc_flags);
-      qla82xx_clear_pending_mbx(vha);
      }
     }
-    ha->flags.isp82xx_fw_hung = 1;
-    ql_log(ql_log_warn, vha, 0xb10a,
-        "Firmware hung.\n");
     qla2xxx_wake_dpc(vha);
    }
   }
-- 
1.7.1